README.md
# community-place-for-cats-discord-bot

[![Python](https://img.shields.io/badge/python-3.10-blue?logo=python)](https://www.python.org/downloads/release/python-3100/)
[![GitHub](https://img.shields.io/github/license/Kotoya1852/community-place-for-cats-discord-bot)](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/blob/main/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/Kotoya1852/community-place-for-cats-discord-bot?logo=github)](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues?q=is%3Aissue+is%3Aopen+)
[![GitHub release (with filter)](https://img.shields.io/github/v/release/Kotoya1852/community-place-for-cats-discord-bot)](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/releases/latest)
[![Static Badge](https://img.shields.io/badge/Visual_Stadio_Code-download-blue?logo=visual-studio-code)](https://azure.microsoft.com/ja-jp/products/visual-studio-code)
[![Static Badge](https://img.shields.io/badge/centos-download-blue?logo=centos)](https://www.centos.org/download/)
[![Static Badge](https://img.shields.io/badge/docker-not_upload-red?logo=docker)]()

> **Warning**<br>
> This document is only available in Japanese.

## 概要
猫の集会場オーナーボット猫のソースコード
[discord.py](https://discordpy.readthedocs.io/ja/latest/) を使用したボット。

## 機能一覧
- 特定サーバーの全ボイスチャンネルの入退室通知
- 先頭に`!`が付いたメッセージをコマンドとして扱った処理の実行
  - ボットのPing値測定結果出力
  - ボットが許容するコマンド一覧出力
  - イベントの参加・辞退
  - イベント情報を出力

## 導入方法
※既に[初回導入](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/wiki/%E5%88%9D%E6%9C%9F%E5%B0%8E%E5%85%A5)を実施済であることが条件です。
1. お好みの場所に移動する
```
cd <お好みのディレクトリ>
```
2. Githubからソースをクローンする
```
git clone https://github.com/Kotoya1852/community-place-for-cats-discord-bot.git -b main
```
3. リポジトリ内に移動する
```
cd community-place-for-cats-discord-bot
```
4. 環境変数を設定するため、set_environment.shを修正する
```
vi set_environment.sh
```
```
export DISCORD_CLIENT_ID=DiscordボットのクライアントID
export DISCORD_NOTIFICATION_CANNEL_ID=入退室通知用テキストチャンネルID
export DISCORD_MINECRAFT_CHANNEL_ID=マインクラフト用テキストチャンネルID
export DISCORD_MINECRAFT_ROLE_ID=マインクラフト参加用ロールID
export DEBUG_LOG_OUTPUT_FLAG=true （※デバッグログを出力する場合のみ、しない場合は`false`を指定しても良い）
export DISCORD_CLIENT_RUN_MODE=normal_run（※メンテナンスモードは `maintenance` を指定する）
```
5. ボットを動かす
```
bash startup.sh
```

## ボットのアップデート方法
※既に導入済であることが前提
1. アップデートスクリプトを実行
```
bash update.sh
```
2. 必要であれば、環境変数を見直します。
```
vi set_environment.sh
```
3. ボットを動かす
```
bash startup.sh
```


CHANGELOG.md
# Changelog
現在の最新バージョンは 2.4.0 です。

## [開発中](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/tree/develop)

### 機能追加

- CHANGELOG.mdを新規作成。

### 仕様変更

### 非推奨対応

### 削除対応

### バグ修正

### セキュリティ対応


## [2.4.0](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/releases/tag/v2.4) - 2023-08-29

### バグ修正

- Dockerfileに記載したプロジェクト名にブランチ名が含まれている [#14](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues/14)
- 入退室通知は全サーバーを見ている [#4](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues/4)
- 入退室通知がされない [#20](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues/20)

## [2.3.0](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/releases/tag/v2.3) - 2023-08-25

### 機能追加

- ボットのDocker化対応（ [#8](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues/8) ）

### 仕様変更

- マインクラフトサーバーのipアドレスがipv6になる（ [#10](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues/10) ）

### セキュリティ対応

- リファクタリング対応（ [#2](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/issues/2) ）


## [2.2.1](https://github.com/Kotoya1852/community-place-for-cats-discord-bot/releases/tag/v2.2) - 2023-08-15

### 仕様変更

- 環境に関するリテラル値を環境変数から取得するよう修正
- Githubで公開し、publicにソースが閲覧できるよう整備
  - これ以降の更新履歴をGithubで管理する。

## 2.2.0 - 2023-08-11

### 機能追加

- イベント参加・辞退コマンドを追加（`!join` / `!leave`）
- イベント情報取得コマンドを追加（`!info`）

### 仕様変更

- ボットステータスのプレゼンス情報をカスタムアクティビティクラスに修正しました。（これ以降、`プレイ中`が削除される）

## 2.1.0 - 2023-07-30

### 機能追加

- コマンドを追加（!help）

### バグ修正

- 入室時に通知が来ないバグを修正


## 2.0.0 - 2023-07-23

### 機能追加

- コマンド機能を追加（!ping）
- ボットステータスを追加（固定文字 `稼働中にゃ！`）

### 仕様変更

- 全ユーザーの通知メッセージがランダムで改変されるようになりました

## 1.3.0 - 2022-08-31

### 機能追加

- 一部ユーザーに対して通知内容が改変されるようになりました

## 1.2.0 - 2022-08-04

### 仕様変更

- 新しくチャンネルを作成しても自動で通知対象になるように修正

## 1.1.0 - 更新日時不明

- 修正内容不明

## 1.0.0 - 2021-08-31

- 新規作成

<!-- Versions -->
[unreleased]: https://github.com/Author/Repository/compare/v0.0.2...HEAD
[0.0.2]: https://github.com/Author/Repository/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/Author/Repository/releases/tag/v0.0.1


discord_channels.py
"""
    機能名：入退室通知ボット猫 - 猫の集会場
    作成者：Kotoya
    作成日：不明
    更新履歴：
        Version 1
        ----------
        v1.0 ; Kotoya ; 2021/08/31; 新規作成
        v1.1 ; Kotoya ; 更新日不明 ; 修正内容不明
        v1.2 ; Kotoya ; 2022/08/04; 新しくチャンネルを作成しても自動で通知対象になるように修正
        v1.3 ; Kotoya ; 2022/08/31; 一部ユーザーに対して通知内容が改変されるようになりました

        Version 2
        ----------
        v2.0 ; Kotoya ; 2023/07/23; 全ユーザーの通知メッセージがランダムで改変されるようになりました
                                    コマンド機能を追加（!ping）
                                    ボットステータスを追加（固定文字 `稼働中にゃ！`）
        v2.1 ; Kotoya ; 2023/07/30; 入室時に通知が来ないバグを修正
                                    コマンドを追加（!help）
        v2.2 ; Kotoya ; 2023/08/11; ボットステータスのプレゼンス情報をカスタムアクティビティクラスに修正しました。（これ以降、`プレイ中`が削除される）
                                    イベント参加・辞退コマンドを追加（!join / !leave）
                                    イベント情報取得コマンドを追加（!info）
        chore; Kotoya ; 2023/08/??; 環境に関するリテラル値を環境変数から取得するよう修正
                                    Githubで公開し、publicにソースが閲覧できるよう整備
                                        → これ以降の更新履歴をGithubで管理する。
"""

import discord
import random
import datetime
import requests
import os

# クライアント
client = discord.Client(intents=discord.Intents.all())
client_id = os.getenv("DISCORD_CLIENT_ID")

# 置換定数
channel_name: str = "$channel_name$"
display_name: str = "$display_name$"

# メッセージリスト
entry_msg_list = list(
    (
        f"**{channel_name}** に、__{display_name}__  が入室しました！",
        f"**{channel_name}** に、__{display_name}__  が入ってきた！にゃおん！",
        f"あ、**{channel_name}** に、__{display_name}__  がきました！",
    )
)

leaving_msg_list = list(
    (
        f"**{channel_name}** から、__{display_name}__  が退出しました！",
        f"**{channel_name}** から、__{display_name}__  が出ていった！にゃおん！",
        f"わーお！**{channel_name}** から、__{display_name}__  がいなくなった！",
    )
)

# テキストチャンネルID
notification_channel_id = os.getenv("DISCORD_NOTIFICATION_CANNEL_ID")
minecraft_channel_id = os.getenv("DISCORD_MINECRAFT_CHANNEL_ID")

# ロールID
minecraft_role_id = os.getenv("DISCORD_MINECRAFT_ROLE_ID")


@client.event
async def on_voice_state_update(
    member: discord.Member, before: discord.VoiceState, after: discord.VoiceState
):
    """
    ボイスチャンネルに入退出したタイミングで発火するイベント処理

    Args:
        member: 入退出したユーザーの情報
        before: 退室元ボイスチャンネルの情報
        after: 入室先ボイスチャンネルの情報
    """
    if before.channel != after.channel:
        # 通知先チャンネル指定
        botRoom = client.get_channel(notification_channel_id)

        # チャンネルIDを自動取得（新しく作成してもソースを修正する必要がない）
        announceChannelIds = []
        for channel in client.get_all_channels():
            announceChannelIds.append(channel.id)

        # デバッグ用-discordにはsendされない-チャンネルIDが入っていることを確認する
        print(announceChannelIds)

        if before.channel is not None and before.channel.id in announceChannelIds:
            # 退出メッセージ送信
            await botRoom.send(random_message_select(leaving_msg_list, before, member))

        if after.channel is not None and after.channel.id in announceChannelIds:
            # 入室メッセージ送信
            await botRoom.send(random_message_select(entry_msg_list, after, member))


@client.event
async def on_message(message: discord.Message):
    """
    コマンドを受け取り、コマンドに見合った処理を実行します。\n
    注意：このイベントはDMでもサーバーの全テキストチャットでも反応します。

    Args:
        message: 受信したメッセージ情報
    """
    print(f"[DEBUG]:{datetime.datetime.now()}:{message.content}")
    if message.content.startswith("!"):
        # コマンド接頭辞は`!`なので、`!`のテキストメッセージを取得する
        # コマンドパラメータは半角スペースで区切る
        input_command_and_parameters = message.content.split(" ")
        input_command = input_command_and_parameters.pop(0)
        if input_command == "!ping":
            # pingコマンド
            # レイテンシーをメッセージに返信する。
            await message.reply(f"pingコマンド実行：ボット猫のPingは約{round(client.latency, 2)}秒です。")
            return
        elif input_command == "!help":
            # helpコマンド
            # 現在使用可能なコマンドをメッセージに返信する。
            commands = list(
                ("`!ping`", "`!join イベント名`", "`!leave イベント名`", "`!info イベント名`")
            )
            commands_help = "\n".join(commands)
            await message.reply(f"helpコマンド実行：現在使用可能なコマンドは以下の通りです。\n{commands_help}")
            return
        elif input_command == "!join":
            # joinコマンド
            # パラメータに指定されたイベントに参加します。
            try:
                event = input_command_and_parameters.pop(0)
                if event == "minecraft":
                    # サーバー情報取得
                    guild = message.guild
                    # ロール情報取得
                    role = discord.utils.get(guild.roles, id=minecraft_role_id)
                    # 特定のロールが付与されているか
                    if message.author.get_role(role.id) is not None:
                        await message.reply(f"既に参加済です。")
                        return
                    # 特定のロールを付与する
                    await message.author.add_roles(role)
                    # 結果を返却する
                    await message.reply(f"イベント`{event}`の参加を受諾しました。")
                    return
                else:
                    # 存在しないイベント
                    await message.reply(f"イベント`{event}`は存在しません。")
            except IndexError:
                # イベントが指定されなかった
                # イベントリストを表示
                events = ["minecraft"]
                event_help = "\n".join(events)
                await message.reply(f"現在参加できるイベントは以下の通りです。\n{event_help}")
                return
        elif input_command == "!leave":
            # leaveコマンド
            # パラメータに指定されたイベントを辞退します。
            try:
                event = input_command_and_parameters.pop(0)
                if event == "minecraft":
                    # サーバー情報取得
                    guild = message.guild
                    # ロール情報取得
                    role = discord.utils.get(guild.roles, id=minecraft_role_id)
                    # 特定のロールが付与されているか
                    if message.author.get_role(role.id) is None:
                        await message.reply(f"イベント`{event}`は参加していません。")
                        return
                    # 特定のロールを除去する
                    await message.author.remove_roles(role)
                    # 結果を返却する
                    await message.reply(f"イベント`{event}`の参加を辞退しました。")
                    return
                else:
                    await message.reply(f"イベント`{event}`は参加していません。")
                    return
            except IndexError:
                # イベントが指定されなかった
                await message.reply("辞退するイベントを指定してください。")
                return
        elif input_command == "!info":
            # infoコマンド
            # パラメータに指定されたイベントの情報を返信します。
            try:
                event = input_command_and_parameters.pop(0)
                if event == "minecraft":
                    # バリデーションチェック
                    # チャンネルIDチェック
                    message_id = message.channel.id
                    if message_id != minecraft_channel_id:
                        # テキストチャンネルIDが一致しない
                        await message.reply(
                            f"このテキストチャンネルでは回答できません。\n`{event}`用のテキストチャンネルでコマンドを実行してください。"
                        )
                        return
                    # ロールmessage.author.get_チェック
                    guild = message.guild
                    member = guild.get_member(message.author.id)
                    print(member)
                    if member == None:
                        # メンバーの情報を取得することに失敗した
                        await message.reply(f"あなたの情報を取得することに失敗しました。管理者に連絡してください。")
                        return
                    role = member.get_role(minecraft_role_id)
                    if role is None:
                        # ロールが見つからない
                        await message.reply(
                            f"あなたに情報を開示する権限がありません。\nイベント`{event}`に参加してください。`!join {event}`"
                        )
                        return

                    # 情報を返信します。
                    ip_addr = get_global_ip_address_by_me()
                    await message.reply(
                        f"【アクセス情報】\nサーバーアドレス：{ip_addr}:25565\nDynmap：{ip_addr}:8123"
                    )
                    return
                else:
                    # 存在しないイベント
                    await message.reply(f"イベント`{event}`について開示できる情報がありません。")
            except IndexError:
                # イベントが指定されなかった
                await message.reply("取得したい情報のイベント名を指定してください。")
                return
        else:
            # 存在しないコマンド
            # エラーメッセージをメッセージに返信する。
            await message.reply(
                f"コマンドリストに `{input_command}` は存在しません。\n`!help`を実行してコマンドを確認してください。"
            )
            return


@client.event
async def on_ready():
    """
    ボットが起動した時に発火するイベントです。
    """
    bot_str = "稼働中にゃ！"
    await client.change_presence(
        status=discord.Status.online,
        activity=discord.Activity(
            name=bot_str,
            type=discord.ActivityType.custom,
            state=bot_str,
            details=bot_str,
        ),
    )

    # デバッグ用-起動した日時をログに残す。
    print(f"{datetime.datetime.now()} : 起動しました。")
    print(f"bot name : {client.user.name}")
    print(f"bot id : {client.user.id}")
    print(f"discord py version : {discord.__version__}")


@client.event
async def on_connect():
    """
    ボットがサーバーに接続した時に発火するイベントです。
    """
    bot_str = "起動中..."
    await client.change_presence(
        status=discord.Status.idle,
        activity=discord.Activity(
            name=bot_str,
            type=discord.ActivityType.custom,
            state=bot_str,
            details=bot_str,
        ),
    )

    # デバッグ用-接続した日時をログに残す。
    print(f"{datetime.datetime.now()} : 正常に接続しました。")


def random_message_select(
    msg_list,
    channel_info: discord.VoiceState,
    member: discord.Member,
) -> str:
    """
    通知メッセージをランダムで生成します。

    Args:
        msg_list: メッセージ候補リスト
        channel_info: チャンネル情報
        member: ユーザー情報
    """
    # 通知メッセージテンプレートをランダムで選択
    msg_len = len(msg_list)
    ran_int = random.randint(0, msg_len - 1)

    # 通知メッセージテンプレートに引数を設定して返却
    message = (
        msg_list[ran_int]
        .replace(channel_name, channel_info.channel.name)
        .replace(display_name, member.display_name)
    )
    print(message)
    return message


def get_global_ip_address_by_me():
    """
    このサーバーのIPアドレスを取得します。
    """
    res = requests.get("https://ifconfig.me")
    return res.text


client.run(client_id)
